### ISSUE TITLE
[data grid] Using Button in Column Header breaks keyboard navigation

### ISSUE THREAD

**Issue by hungmanhle** [mui/mui-x#18876]
### Search keywords

header, button, accessibility, a11y, keyboard navigation, renderHeader

### Latest version

- [x] I have tested the latest version

### Affected products

Data Grid

### Steps to reproduce

Steps:
1. Open this link to live example: https://codesandbox.io/p/devbox/twilight-voice-mpgc9x?file=%2Fsrc%2FDemo.tsx%3A36%2C4&from-embed=&workspaceId=ws_RKWgwxRpTWheGpvQ3ZG93b
2. Use Tab to navigate into one of the column headers.
3. Attempt to navigate data grid using keyboard.

### Current behavior

"Grid component" keyboard navigation no longer works as expected. Arrow keys do not move between cells. Tab will take you to the next button element instead of taking you out of the data grid.

### Expected behavior

Expect that using button elements in each column header does not effect keyboard navigation.

### Context

We are currently under accessibility audit under LevelAccess and was told that each column header, since it is a clickable element that sorts the table, should have a button role.

After trying to include buttons by using the `renderHeader` prop, I noticed that keyboard navigation no longer works as expected.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>
  System:
    OS: macOS 15.5
  Binaries:
    Node: 22.17.1 - ~/.nvm/versions/node/v22.17.1/bin/node
    npm: 10.9.2 - ~/.nvm/versions/node/v22.17.1/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 138.0.7204.158
    Edge: Not Found
    Safari: 18.5
  npmPackages:
    @emotion/react:  11.14.0
    @emotion/styled:  11.14.1
    @mui/base:  5.0.0-beta.70
    @mui/core-downloads-tracker:  6.4.11
    @mui/icons-material:  6.1.6
    @mui/lab:  6.0.0-beta.31
    @mui/material:  6.4.11
    @mui/private-theming:  6.4.9
    @mui/styled-engine:  6.4.11
    @mui/system:  6.4.11
    @mui/types:  7.2.24
    @mui/utils:  7.0.2
    @mui/x-data-grid:  7.29.1
    @mui/x-data-grid-pro:  7.29.1
    @mui/x-date-pickers:  7.29.1
    @mui/x-date-pickers-pro:  7.29.1
    @mui/x-internals:  7.29.0
    @mui/x-license:  7.29.1
    @mui/x-tree-view:  7.29.1
    @types/react: 18.3.23 => 18.3.23
    react: 18.3.1 => 18.3.1
    react-dom: 18.3.1 => 18.3.1
    typescript: 5.8.3 => 5.8.3
```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by michelengelen** [comment]
simply setting `tabIndex={-1}` on the Button component fixes the keyboard navigation.
It does not break other functionality. You can still activate the button on hitting Enter.
Does this solve your issue?

**Comment by hungmanhle** [comment]
This does not quite resolve the issue because, although keyboard navigation is fixed, the buttons are not being focused and therefore aren't announced to screenreaders.

<img width="788" height="805" alt="Image" src="https://github.com/user-attachments/assets/815a2622-be19-4120-99a4-3850943eb4a9" />

**Comment by michelengelen** [comment]
> This does not quite resolve the issue because, although keyboard navigation is fixed, the buttons are not being focused and therefore aren't announced to screenreaders.
> 
> <img alt="Image" width="788" height="805" src="https://private-user-images.githubusercontent.com/6510778/469852883-815a2622-be19-4120-99a4-3850943eb4a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3OTE4MjYsIm5iZiI6MTc1Mzc5MTUyNiwicGF0aCI6Ii82NTEwNzc4LzQ2OTg1Mjg4My04MTVhMjYyMi1iZTE5LTQxMjAtOTlhNC0zODUwOTQzZWI0YTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MjlUMTIxODQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzE3YTFmYjAyNzllYzZmMzFmMjdjOGFkYmE2NGE3MWNhMjBiNzA1Y2JlYzBiNThmOTZiNjMyY2Q2YWVmMDNmNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FqluUX1X7pqJ9MyAX8pCvgSHMYJX4cwslE5qhhSBfqU">

thats true, but wouldn't setting `role='button'` fix this?

**Comment by hungmanhle** [comment]
@michelengelen Setting the `role='button'` does not fix this. Since we've set the buttons to have `tabIndex={-1}` I assume the button element is never focused, only the outer div with `role='columnheader'` is focused. 
If you mean to set the outer div to `role='button'`, I assume that would fix this but I'm not sure how I would access it. The `renderHeader` only adds markup that to inside the `columnheader`.

<img width="875" height="875" alt="Image" src="https://github.com/user-attachments/assets/d5a4ace5-bb1d-42bb-b938-07537bf555aa" />

<img width="1332" height="220" alt="Image" src="https://github.com/user-attachments/assets/051e8267-9a1c-4f05-8150-7976b369a5a9" />

**Comment by michelengelen** [comment]
Yes, it's not possible to access the rendering of the header container. I am curious though: none of the other grids we know use the `role='button'` prop.

I did a bit of research and there is no clear need for the an interactive role, like `"button"`, on a column header. It is described in the w3c draft as a subclass o `"gridcell"` and `"cell"` and both of them are defined as interactive. Reference: https://w3c.github.io/aria/#columnheader

So we are actually defining the right values here and this should pass the audit without problems when rendering the default columheader. If not please post any report (ideally with a reference) here why it did not pass and we will look into it.

cc @cherniavskii

**Comment by hungmanhle** [comment]
I'm not able to post the actual audit report but the reference they gave me for what the expected behavior was, is this sample from w3c: https://www.w3.org/WAI/ARIA/apg/patterns/grid/examples/data-grids/

Specifically example 2

<img width="925" height="933" alt="Image" src="https://github.com/user-attachments/assets/bf8de397-b033-4cf7-9849-b9f3879c15bc" />

**Comment by michelengelen** [comment]
all right ... it is indeed a bit different in our case.

The differences being:
1. the click handler is registered on the `GridColumnHeaderItem` component: https://github.com/mui/mui-x/blob/78c5912a89d746d704d5590af3ffa8bfc1a900af/packages/x-data-grid/src/components/columnHeaders/GridColumnHeaderItem.tsx#L168

This should ideally be handled on the `ColumnHeaderInnerTitle` component instead to align with other implementations and allow for correct a11y features.

2. The `role='button'` property should be set based on the ability to sort the column

I don't think that this is an easy task to accomplish and it will definitely be a breaking change, so adjusting this might take a while.

I'll add it to the board for the team to align on this and estimate it.
Maybe there is a way to achieve this I am currently missing here that someone else knows. 🤷🏼

**Comment by hungmanhle** [comment]
Hi @michelengelen I've been playing with this for a while and ~I found almost a workaround. Please see the original sandbox which has been edited: https://codesandbox.io/p/devbox/twilight-voice-mpgc9x?file=%2Fsrc%2FDemo.tsx%3A21%2C1&from-embed=&workspaceId=ws_RKWgwxRpTWheGpvQ3ZG93b~

~Giving the button an onKeyDown handler that forwards the event up to the `columnheader` fixes the keyboard navigation using the arrow keys.~

~The only issue here is that tabindexing is incorrect for the buttons. They are always set to `tabIndex={0}`.~

~You would be able to tab into each button. The `renderHeader` method would need to pass down the `tabIndex` of the `columnheader` to fix this. This would also align it with the parameters passed by the `renderCell` method.~

Turns out the solution I had still had many issues.
